## ЧАСТЬ I. Решите задания используя только стандартные методы Python 3

### 1. Импортируйте число e из библиотеки math и выведите его с окргулением до 4 цифр после запятой

In [204]:
import numpy as np
from math import e
from math import sqrt


print(round(e, 4))

2.7183


### 2. Вычислите целую часть результата деления 100 на 7. Убедитесь, что её тип – целое число.

In [205]:
result = 100 // 7

print(result)

if (type(result) == int):
    print('result is integer')

14
result is integer


### 3. Создайте список с квадратами целых чисел от 12 до 4 с шагом 2 не делящимися на 5.

In [206]:
squares = [i**2 for i in reversed(range(4, 13, 2)) if i**2 % 5 != 0]
print(squares)

[144, 64, 36, 16]


### 4.	Отсортируйте полученный в пункте 3 список относительно модуля разницы между элементами этого списка и числом 50.

In [207]:
squares.sort(key = lambda x: abs(x - 50))
print(squares)

[64, 36, 16, 144]


### 5. Создайте множество (set) содержащее все 26 маленькие буквы латинского алфавита.

In [208]:
alphabet = set(map(chr, range(ord('a'), ord('z') + 1))) # all lowercase letters from a to z
print(alphabet)
print(len(alphabet))

{'n', 'u', 'b', 'w', 'y', 'x', 'm', 'v', 'd', 'q', 'k', 'z', 'l', 'i', 'h', 'f', 'a', 'j', 'c', 'p', 'e', 't', 's', 'o', 'r', 'g'}
26


### 6.	Создайте список, содержащий числа с повторениями. Используя словарь, подсчитайте, сколько раз каждое число встречалось в этом списке, и напечатайте содержимое словаря

In [209]:
numbers = [i % 15 for i in range(100)]

occurrences = {}

for key in numbers:
    occurrences[key] = occurrences.get(key, 0) + 1

print(occurrences)

{0: 7, 1: 7, 2: 7, 3: 7, 4: 7, 5: 7, 6: 7, 7: 7, 8: 7, 9: 7, 10: 6, 11: 6, 12: 6, 13: 6, 14: 6}


### 7. Реализуйте функцию от трёх аргументов a, b и c находящую вещественные корни уравнения ax2 + bx + c = 0, с нулевыми значениями аргументов по умолчанию

In [210]:
def get_roots(a = 0, b = 0, c = 0):
    discriminant = b**2 - 4*a*c
    x = None
    y = None

    if (discriminant >= 0):
        x1 = (-b + sqrt(discriminant)) / (2*a)
        x2 = (-b - sqrt(discriminant)) / (2*a)

    return (x1, x2)

print(get_roots(3, -14, -5))

(5.0, -0.3333333333333333)


### 8.	Реализуйте класс Point с полями x,y заполняемыми в конструкторе

In [211]:
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def calc_euclid_dist(self, point):
        return sqrt((self.x - point.x)**2 + (self.y - point.y)**2)


point_inside = Point(4, 10)
point_outside = Point(5, 12)

### 9. Реализуйте класс Circle наследника класса Point с дополнительным полем r и методом inside(p) проверяющим, что круг одержит точку p.

In [212]:
class Circle(Point):
    def __init__(self, x, y, radius):
        super().__init__(x, y)
        self.r = radius

    def inside(self, p: Point):
        return ((p.x - self.x)**2 + (p.y - self.y)**2) <= self.r**2


circle = Circle(1, 2, 9)
print(circle.inside(point_inside))
print(circle.inside(point_outside))

True
False


### 10. Реализуйте метод класса Point, принимающий другой объект класса Point и вычисляющий эвклидово расстояние до него

In [213]:
print(point_outside.calc_euclid_dist(point_inside))

2.23606797749979


## ЧАСТЬ II. Решите задачи используя библиотеку NumPy

### 11. Создайте матрицу 4x4, значения элементов которой варьируются от 0 до 15

In [214]:
m44 = np.arange(16).reshape(4, 4)
print(m44)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]


### 12. Найдите индексы ненулевых элементов из списка [1,2,0,0,4,0,5,7,8,0,0,0]

In [215]:
lst = np.array([1, 2, 0, 0, 4, 0, 5, 7, 8, 0, 0, 0])

indices = np.where(lst != 0)
print(indices)

(array([0, 1, 4, 6, 7, 8]),)


### 13. Создайте единичную матрицу размера 3х3

In [216]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

### 14. Создайте матрицу 10х10, заполните её случайными значениями и найдите максимальный и минимальный элементы

In [217]:
m1010 = np.random.random((10, 10))
m1010_min = m1010.min()
m1010_max = m1010.max()

print(m1010)
print(m1010_min, m1010_max)

[[0.77782426 0.32266823 0.76688332 0.91162849 0.87987166 0.01414243
  0.85415608 0.27729725 0.26718599 0.69202165]
 [0.62669816 0.70568113 0.19160035 0.04683901 0.74827678 0.25434951
  0.73977461 0.50433172 0.37936261 0.69003317]
 [0.46807896 0.95966674 0.95754687 0.26467796 0.94192178 0.99345412
  0.50309214 0.77507936 0.08406694 0.03321164]
 [0.77937931 0.20668161 0.25369378 0.48663281 0.59003858 0.89762417
  0.34052982 0.64755204 0.39496703 0.8617031 ]
 [0.93809816 0.36341159 0.85954366 0.51130506 0.99184836 0.26807027
  0.94838913 0.75342376 0.45533957 0.87794124]
 [0.6821038  0.12417433 0.4999044  0.18218804 0.53300411 0.71048059
  0.28701215 0.2066452  0.17180196 0.56354783]
 [0.81854615 0.48478327 0.55825424 0.40858753 0.19793679 0.60521917
  0.43081813 0.64475534 0.73056228 0.05386615]
 [0.2457814  0.81149304 0.07436877 0.85335971 0.97262788 0.52348676
  0.66355566 0.62686886 0.86686631 0.48081289]
 [0.48237266 0.71619523 0.93185589 0.55092116 0.15066206 0.41643776
  0.07228571

### 15. Создайте матрицу 5х5 и расположите в ней числа 1, 2, 3, 4 под главной диагональю

In [218]:
m55 = np.diag(np.arange(1, 5), k = -1)
print(m55)

[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]


### 16. Перемножьте случайные матрицы размерностей 5х3 и 3х2

In [219]:
m53 = np.random.random((5, 3))
m32 = np.random.random((3, 2))

print(np.dot(m53, m32))

[[0.74256207 1.40694212]
 [0.47844735 0.73348358]
 [0.61741728 1.28640918]
 [1.05199003 1.69183946]
 [0.44902411 0.87697048]]


### 17. Найдите число из списка t, наиболее близкое по величине к числу a

In [220]:
a = 2.39
t = [0, 1, 1.2, 2, 3, 9]

new_t = np.abs(np.array(t) - a)
index = new_t.argmin()
result = t[index]

print(result)

2


### 18. Создайте массив из 10 элементов и поменяйте значение его наибольшего элемента на 0 

In [221]:
m_max = np.random.random((10))
m_max[m_max.argmax()] = 0
print(m_max)

[0.63139903 0.30141642 0.9650559  0.         0.21867009 0.6561207
 0.86397509 0.32955182 0.91765823 0.47211252]


### 19. Создайте матрицу 5x5 и поменяйте в ней местами 2 и 3 строку

In [222]:
matrix = np.random.randint(0, 9, size = (5, 5))
print(matrix)
print('\n')

matrix[2], matrix[3] = matrix[3], matrix[2].copy()
print(matrix)

[[8 0 7 8 0]
 [1 1 1 0 7]
 [1 5 6 7 4]
 [5 1 0 0 7]
 [2 0 7 2 3]]


[[8 0 7 8 0]
 [1 1 1 0 7]
 [5 1 0 0 7]
 [1 5 6 7 4]
 [2 0 7 2 3]]


### 20. Создайте случайную целочисленную матрицу и найдите наиболее часто встречающийся в ней элемент

In [223]:
n = np.random.randint(2, 20)
m = np.random.randint(2, 20)

matrix = np.random.randint(0, 100, size = (n, m))
print(matrix)

unique = np.unique(matrix, return_counts=True)
index = unique[1].argmax()
result = unique[0][index]

print(result)

[[44 25 60 19 91 52 77 43 66]
 [95 53 60 84 77 63 54 20 96]
 [69 68 91 76 16 20  7 87 89]
 [25 37 55 78 73 99 86 29 78]
 [75 92  9  9 84 76 33 72 96]
 [87 55 66 70 66 19 51 20 40]
 [96 94 47 59 59 99 77 82 30]
 [95 35 55 41 16 25 97 31 40]
 [53 82 69 67 68 48 70 70  8]
 [ 7  7  1 51 71 12 42 63  8]
 [36 67 37 35 29 91 65 89 36]
 [ 3 93 92 92  5 62 37 58 88]
 [12 51 15 29 90 51 84  6 67]
 [56 23 78 57 64 48 34 46 18]
 [41 89  1 33 34 59 23 71 79]]
51


## Часть III. Прочитайте краткий мануал по библиотеке pandas и выполните оставшиеся задания

Pandas - это библиотека Python, предоставляющая широкие возможности для анализа данных. С ее помощью очень удобно загружать, обрабатывать и анализировать табличные данные с помощью SQL-подобных запросов. В связке с библиотеками Matplotlib и Seaborn появляется возможность удобного визуального анализа табличных данных.


Для использования библиотек их необходимо импортировать, что можно сделать с помощью команды import. После ключевого слова import указывается название библиотеки, а после as - ее краткое название (алиас), которое мы будем использовать в коде для краткого обозначения этой библиотеки.

In [224]:
import pandas as pd

Основными структурами данных в Pandas являются классы Series и DataFrame. Первый из них представляет собой одномерный индексированный массив данных некоторого фиксированного типа (его можно интеерпретировать как индексированный столбец таблицы). Второй - это двухмерная структура данных, представляющая собой таблицу, каждый столбец которой содержит данные одного типа. Можно представлять её как словарь объектов типа Series. Структура DataFrame отлично подходит для представления реальных данных: строки соответствуют признаковым описаниям отдельных объектов, а столбцы соответствуют признакам.

## Чтение из файла и первичный анализ
Чаще всего, данные, которые мы будем анализировать, готовить  и использовать для обучения моделей, не формируются самостоятельно, а берутся из готового дататсета. Рассмотрим чтение датасета (DataFrame) из файла с расширением csv. В качестве примера возьмем titanic_train.csv, который содержить данные о пассажирах Титаника. Полный датасет titanic был разделен на 2 части - тренировочную и тестовую. Именно с тренировочной частью мы и будем работать. Датасет titanic предоставлен в рабочей папке Лаб_1, а также вы можете его скачать (titanic_train.csv) по ссылке https://www.kaggle.com/tedllh/titanic-train#titanic_train.csv. Если вы работаете на платформе Goggle Colab, необходимо сделать upload файла в вашу рабочую директорию ( в меню ноутбука слева  найти и перейти на Files, нажать Upload и выбрать файл для загрузки)

Внимание! Проверьте, что файл находиться в папке content, если его там нет, то нажмите на нее правой кнопкой и upload. 


### Чтение данных

Для чтения csv файла используем функцию pd.read_csv

In [225]:
import pandas as pd
df = pd.read_csv('titanic_train.csv')

Посмотрим тип переменной df

In [226]:
type(df)

pandas.core.frame.DataFrame

### Первичный анализ

После того, как мы считали данные, посмотрим первые несколько записей датафрейма (если в скобках после функции head не указать нужное количество записей, то по умолчанию будут выведены первые 5 записей датафрейма). Если вместо таблицы вы видите  "кашу" из данных, разделенных ";", значит, ваш csv файл использует в качестве разделителя ";", а не ",". И для корректного чтения данных, нужно задать разделитель в функции чтения, например df = pd.read_csv('proba.csv', sep=";")

In [227]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Давайте посмотрим, что это за данные. Достаточно часто датасеты имеют краткое описание содержимого. Например, что означает тот или иной столбец, какие значения используются для заполнения и т.д. Или значения признаков интуитивно понятны. В нашем случае, мы имеем следующие данные:
каждая строка - это сведения об отдельном пассажире. Строка содержит:
*   PassengerId - идентификатор пассажира (проще говоря, порядковый номер строки)
*   Survived - признак, выжил ли пассажир после крушения (0 - погиб, 1 - выжил)
*   Pclass - в каком классе ехал пассажир (1, 2 или 3)
*   Name - имя пассажира
*   Sex - пол (male - мужской, female - женский)
*   Age - возраст
*   SibSp - количество родственников (по горизонтали - супруг, братья, сестры) на борту
*   Parch - количество родственников (по вертикали - родители, дети) на борту 
*   Ticket - номер билета
*   Fare - цена билета
*   Cabin - номер каюты
*   Embarked - порт посадки (C — Cherbourg, Q — Queenstown, S — Southampton)





Команда shape() показывате размерность датафрейма

Мы видим, что в нашем датафрейме 891 строка (то есть, сведения о 891 пассажире) и 12 столбцов признаков

In [228]:
df.shape

(891, 12)

Команда columns.values() показывате название всех признаков

In [229]:
df.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

Также полезно посмотреть метод info, который показывает краткую статистику по набору данных

In [230]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Можно увидеть, что не все признаки у нас заполнены в каждой строке,для некоторых признаков есть пропуски. Например, у нас 891 строка, а Age (возраст) заполнен только в 714 строках. Совсем мало значений в столбце Cabin. 

Есть несколько вариантов работы с такими строками: 
* их можно оставить без изменений, если признак, в котором пропущены данные, малозначим;
* можно удалить все строки, в которых есть пустые значения;
* можно заполнить их какими-либо значениями (0, медианой или любыми другими константами).

Метод info так же показывает какого типа данные в том или ином столбце: цифровые (int или float), текстовые и т.д. В последней строке указан объем памяти, которую занимает датафрейм. При больших объемах данных это позволяет оценить хватит ли вам ресурсов для загрузки в память полного датасета

Еще один полезный метод для начального анализа данных - метод describe(), который выдает краткую статистику по числовым признакам: среднее, стандартное отклонение, минимум, максимум, медианы, квартили и т.д. Самостоятельно посмотрите, что означает каждое понятие. 

In [231]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Доступ к элементам данных

### Доступ к элементам DataFrame

Разберемся как посмотреть какие-то конкретные значения в датафрейме. Наш датафрейм представляет собой таблицу, состоящую из строк и столбцов. Обратиться к конкретному столбцу можно используя его имя. Например, выведем первые 10 строк из столбца с возрастом пассажиров

In [232]:
df['Age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Второй способ обращения - через "."

In [233]:
df.Age.head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Столбец таблицы - это объек Pandas Series, убедимся в этом определив тип полученного объекта. Этот объект проиндексирован (слева печатаются индексы). В нижней строке есть информацию о его имени и типе значений.
Кроме того, здесь хорошо видны пропуски (незаполненные значения) в некоторых строках (они обозначены NaN - Not-A-Number)

In [234]:
type(df['Age'])

pandas.core.series.Series

Кроме обращения к столбцу по имени, можно использовать индексацию с помощью методов loc и iloc

Метод loc выдает нам подматрицу, используя индексы строк и столбцов. Первый аргумент отвечает за количество записей (как с начала, так и при желании с любого места), а второй - за отраженные в подматрице признаки.


In [235]:
df.loc[0:5, ['Pclass', 'Fare']]

Pclass     Fare
0       3   7.2500
1       1  71.2833
2       3   7.9250
3       1  53.1000
4       3   8.0500
5       3   8.4583

Метод iloc работает схожим образом, за исключением того, что второй аргумент теперь тоже индексы, то есть отражает количество признаков.


In [236]:
df.iloc[0:5, 1:3]

Survived  Pclass
0         0       3
1         1       1
2         1       3
3         1       1
4         0       3

# Доступ к элементам Series

Для примера создадим два объекта Series

In [237]:
s1 = pd.Series(["one", "two", "three", "four", "five"]) # передаем список, индексы проставляются автоматически с 0
print(s1)

0      one
1      two
2    three
3     four
4     five
dtype: object


In [238]:
s2 = pd.Series([1, 2, 3, 4, 5], ['a', 'b', 'c', 'd', 'e'])  # передаем список, индексы заданы символами от 'a' до 'e'
print(s2)

a    1
b    2
c    3
d    4
e    5
dtype: int64


К элементам Series можно обращаться по численному индексу через "[ ]"

In [239]:
s2[2] # второй элемент серии s2 (индексация с 0)

3

К элементам Series можно обращаться по метке:

In [240]:
s2['c']

3

К элементам Series можно применять так называемые slice (срезы) через ':'.
Срезами является часть серии до или после определенного элемента

In [241]:
s2[:2] # все элементы до второго включительно

a    1
b    2
dtype: int64

In [242]:
s2[2:] # второй и все элементы после второго до конца

c    3
d    4
e    5
dtype: int64

In [243]:
s2[2:4] # все элементы со второго по четвертый не включительно. Последний никогда не включается.

c    3
d    4
dtype: int64

In [244]:
s2[:] # все элементы

a    1
b    2
c    3
d    4
e    5
dtype: int64

В поле для индекса можно поместить условное выражение.
Данное выражение можно перевести как вопрос, является ли значение элемента (признака) меньшим или равным двум.

In [245]:
s2[s2 <= 2]

a    1
b    2
dtype: int64

Или даже несколько условных выражений, отделенных "( )". 

Если соединить два условия знаком "&", то результат будет удовлетворять двум условиям, а если знаком "|", то хотя бы одному из них.

In [246]:
s2[(s2 >= 2) & (s2 <=4)]

b    2
c    3
d    4
dtype: int64

In [247]:
s2[(s2 <= 2) | (s2 == 5)]

a    1
b    2
e    5
dtype: int64

Важно, что в условных выражениях равенство пишется как "==" (два знака "равно").

Со структурами Series можно работать как с векторами: складывать, умножать вектор на число и т.п.

In [248]:
s3 = s2 + 10

In [249]:
print(s3)

a    11
b    12
c    13
d    14
e    15
dtype: int64


## Примеры анализа информации

Для примера отберем пассажиров мужского пола, возраст которых меньше сорока лет.

In [250]:
df[(df['Sex'] == 'male') & (df.Age < 40)].head(5)

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0             1         0       3         Braund, Mr. Owen Harris  male  22.0   
4             5         0       3        Allen, Mr. William Henry  male  35.0   
7             8         0       3  Palsson, Master. Gosta Leonard  male   2.0   
12           13         0       3  Saundercock, Mr. William Henry  male  20.0   
13           14         0       3     Andersson, Mr. Anders Johan  male  39.0   

    SibSp  Parch     Ticket    Fare Cabin Embarked  
0       1      0  A/5 21171   7.250   NaN        S  
4       0      0     373450   8.050   NaN        S  
7       3      1     349909  21.075   NaN        S  
12      0      0  A/5. 2151   8.050   NaN        S  
13      1      5     347082  31.275   NaN        S

Можно отсортировать этих людей по возрастанию платы за билет. Ascending в данной функции отвечает за сортировку по убыванию (False) или по возрастанию (True).

In [251]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).head(10)

PassengerId  Survived  Pclass                             Name   Sex  \
271          272         1       3     Tornquist, Mr. William Henry  male   
822          823         0       1  Reuchlin, Jonkheer. John George  male   
806          807         0       1           Andrews, Mr. Thomas Jr  male   
179          180         0       3              Leonard, Mr. Lionel  male   
302          303         0       3  Johnson, Mr. William Cahoone Jr  male   
378          379         0       3              Betros, Mr. Tannous  male   
872          873         0       1         Carlsson, Mr. Frans Olof  male   
843          844         0       3       Lemberopolous, Mr. Peter L  male   
371          372         0       3        Wiklund, Mr. Jakob Alfred  male   
202          203         0       3       Johanson, Mr. Jakob Alfred  male   

      Age  SibSp  Parch   Ticket    Fare        Cabin Embarked  
271  25.0      0      0     LINE  0.0000          NaN        S  
822  38.0      0      0    19972  0.0000          NaN        S  
806  39.0      0      0   112050  0.0000          A36        S  
179  36.0      0      0     LINE  0.0000          NaN        S  
302  19.0      0      0     LINE  0.0000          NaN        S  
378  20.0      0      0     2648  4.0125          NaN        C  
872  33.0      0      0      695  5.0000  B51 B53 B55        S  
843  34.5      0      0     2683  6.4375          NaN        C  
371  18.0      1      0  3101267  6.4958          NaN        S  
202  34.0      0      0  3101264  6.4958          NaN        S

Можно подсчитать количество пассажиров (а точнее их уникальных номеров) с помощью функции count()

In [252]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).PassengerId.count()

344

Также можно быстро вывести распределение переменной Sex с помощью функции value_count()

In [253]:
df.Sex.value_counts()

male      577
female    314
Name: Sex, dtype: int64

## Задания

### 21. Сколько человек спаслись и сколько погибли?

In [261]:
print('Выжило:', df[df.Survived == 1].Survived.count())
print('Погибло:', df[df.Survived == 0].Survived.count())

Выжило: 342
Погибло: 549


### 22. Выведите распределение переменной Survived отдельно для мужчин и отдельно для женщин


In [269]:
print('Мужчины:\n', df[df.Sex == 'male'].Survived.value_counts())
print('\n')
print('Женщины:\n', df[df.Sex == 'female'].Survived.value_counts())

Мужчины:
 0    468
1    109
Name: Survived, dtype: int64


Женщины:
 1    233
0     81
Name: Survived, dtype: int64


### 23. Каковы медиана и стандартное отклонение цен за билет?

In [275]:
df.Fare.describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

### 24. Подсчитайте долю выживших среди людей моложе 25 лет и старше 55.

In [297]:
old_n_young = df[(df.Age < 25) | (df.Age >= 55)]
percent = old_n_young[old_n_young.Survived == 1].PassengerId.count() / old_n_young[old_n_young.Survived == 0].PassengerId.count()
print(percent)

0.6931216931216931


### 25. Подсчитайте долю выживших среди мужчин и женщин.

In [308]:
men = df[(df.Sex == 'male')]
men_survived = men[(men.Survived == 1)].PassengerId.count()

women = df[(df.Sex == 'female')]
women_survived = women[(women.Survived == 1)].PassengerId.count()

print(men_survived / men.PassengerId.count())
print(women_survived / women.PassengerId.count())

0.18890814558058924
0.7420382165605095


### 26. В каком городе больше всего село пассажиров?

In [315]:
df.Embarked.value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

### 27. Подсчитайте средний возраст выживших

In [318]:
df[df.Survived == 1].Age.mean()

28.343689655172415

### 28. Подсчитайте средний возраст пассажиров каждого из трех классов


In [323]:
class1 = df[df.Pclass == 1].Age.mean()
class2 = df[df.Pclass == 2].Age.mean()
class3 = df[df.Pclass == 3].Age.mean()

print(class1, class2, class3)

38.233440860215055 29.87763005780347 25.14061971830986


### 29. Подсчитайте средний возраст выживших мужчин, купивших билет во 2 класс

In [326]:
df[(df.Sex == 'male') & (df.Survived == 1) & (df.Pclass == 2)].PassengerId.count()

17

### 30. Оцените, как влияет на выживаемость пассажиров наличие их родственников на борту.

In [332]:
print(df[(df.SibSp > 0) | (df.Parch > 0)].Survived.value_counts())
print('\n')
print(df[(df.SibSp == 0) & (df.Parch == 0)].Survived.value_counts())

1    179
0    175
Name: Survived, dtype: int64


0    374
1    163
Name: Survived, dtype: int64
